File: leet-code/maxNumber.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number[]}
 */
var maxNumber = function (nums1, nums2, k) {
  let l1 = nums1.length
  let l2 = nums2.length
  var getMaxNumbers = function (nums, k) {
    const n = nums.length
    const stack = []
    for (let i = 0; i < n; i++) {
      while (k > 0 && stack.length > 0 && stack[stack.length - 1] < nums[i]) {
        stack.pop()
        k--
      }
      stack.push(nums[i])
    }
    k > 0 ? stack.splice(-k) : stack
    return stack
  }

  let res = []
  for (let i = 0; i <= k; i++) {
    rest1 = l1 - i
    rest2 = l2 - k + i
    if (rest1 >= 0 && rest2 >= 0 && rest1 <= l1 && rest2 <= l2) {
      res = arrmax(
        res,
        merge(getMaxNumbers(nums1, l1 - i), getMaxNumbers(nums2, l2 - k + i))
      )
    }
  }
  return res
}

function arrmax (arr1, arr2) {
  const l1 = arr1.length
  const l2 = arr2.length
  if (l1 > l2) {
    return arr1
  }
  if (l1 < l2) {
    return arr2
  }
  for (let i = 0; i < l1; i++) {
    if (arr1[i] > arr2[i]) {
      return arr1
    }
    if (arr2[i] > arr1[i]) {
      return arr2
    }
  }
  // 全部相等,随便返回一个
  return arr1
}

function merge (subsequence1, subsequence2) {
  const x = subsequence1.length,
    y = subsequence2.length
  if (x === 0) {
    return subsequence2
  }
  if (y === 0) {
    return subsequence1
  }
  const mergeLength = x + y
  const merged = new Array(mergeLength).fill(0)
  let index1 = 0,
    index2 = 0
  for (let i = 0; i < mergeLength; i++) {
    if (compare(subsequence1, index1, subsequence2, index2) > 0) {
      merged[i] = subsequence1[index1++]
    } else {
      merged[i] = subsequence2[index2++]
    }
  }
  return merged
}

function compare (subsequence1, index1, subsequence2, index2) {
  const x = subsequence1.length,
    y = subsequence2.length
  while (index1 < x && index2 < y) {
    const difference = subsequence1[index1] - subsequence2[index2]
    if (difference !== 0) {
      return difference
    }
    index1++
    index2++
  }
  return x - index1 - (y - index2)
}

console.log(maxNumber([3, 4, 6, 5], [9, 1, 2, 5, 8, 3], 5))
console.log(maxNumber([6, 7], [6, 0, 4], 5))
console.log(maxNumber([3, 9], [8, 9], 3))
console.log(maxNumber([8, 6, 9], [1, 7, 5], 3))
console.log(maxNumber([6, 7], [6, 0, 4], 5))

File: leet-code/maxNumber.go
package main

import "fmt"

func maxNumber(nums1 []int, nums2 []int, k int) []int {
	var ans []int
	l1, l2 := len(nums1), len(nums2)
	for i := 0; i <= k; i++ {
		rest1 := l1 - i
		rest2 := l2 - (k - i)
		ans = arrmax(ans, merge(findMaxNum(nums1, rest1), findMaxNum(nums2, rest2)))
	}
	return ans
}

// nums 数组
// k 要删除的数量
func findMaxNum(nums []int, k int) []int {
	if len(nums) == 0 {
		return nums
	}
	stack := []int{nums[0]}
	for i := 1; i < len(nums); i++ {
		for k > 0 && len(stack) > 0 && stack[len(stack)-1] < nums[i] {
			stack = stack[0 : len(stack)-1]
			k--
		}
		stack = append(stack, nums[i])
	}
	if k > 0 {
		stack = stack[0 : len(stack)-k]
	}
	fmt.Println(stack)
	return stack
}

func merge(n1, n2 []int) []int {
	l1, l2 := len(n1), len(n2)
	if l1 == 0 {
		return n2
	}
	if l2 == 0 {
		return n1
	}
	mergeLen := l1 + l2
	merged := make([]int, mergeLen)
	var idx1, idx2 int
	for i := 0; i < mergeLen; i++ {
		if compare(n1, n2, idx1, idx2) {
			merged[i] = n1[idx1]
			idx1++
		} else {
			merged[i] = n2[idx2]
			idx2++
		}
	}
	return merged
}

func compare(n1, n2 []int, idx1, idx2 int) bool {
	l1, l2 := len(n1), len(n2)
	for idx1 < l1 && idx2 < l2 {
		diff := n1[idx1] - n2[idx2]
		if diff != 0 {
			return diff > 0
		}
		idx1++
		idx2++
	}
	return l1-idx1-(l2-idx2) > 0
}

func arrmax(n1, n2 []int) []int {
	l1, l2 := len(n1), len(n2)
	if l1 > l2 {
		return n1
	} else if l2 > l1 {
		return n2
	} else {
		for i := 0; i < l1; i++ {
			if n1[i] > n2[i] {
				return n1
			} else if n1[i] < n2[i] {
				return n2
			}
		}
	}
	return n1
}
